File: benchmarks/python_cost/timer.cpp
// The C++ side of this capability's own Python-vs-C++ per-callback cost
// comparison: loops the same evaluate() call many times and times it, for
// two pi shapes (trivial, real Python-side work) crossed with two sampling
// configurations (replenishing, non-replenishing -- see this file's own
// module doxygen below for why both). timer.py is the identical
// comparison from the Python binding's own side; report.py runs both and
// tabulates the ratio.
//
// Not a cc_test: this measures wall time, which is exactly what
// benchmarks/belief_evaluation/instrument.cpp's own header comment already
// explains a cc_test must never assert on. A cc_binary instead, not tagged
// manual, so `bazel test //...` builds it everywhere and never runs it --
// running it is report.py's job.
//
// Solver-free: nothing here needs solve_board. The pi/delta pair timed
// below lives in strategies.hpp, beside this file -- see that header's own
// comment for why it is not library/tests/belief_evaluation/test_support.hpp
// (this plan's own scope rule: no executable line changed under
// `library/`).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>

#include <api/dds_constants.hpp>
#include <api/dds_data_types.hpp>

#include <belief_evaluation/declarer_strategy.hpp>
#include <belief_evaluation/evaluate.hpp>
#include <belief_evaluation/exhaustive_layout_source.hpp>

#include "strategies.hpp"

namespace be = dds::belief_evaluation;
namespace strategies = dds::belief_evaluation::benchmarks::python_cost;

namespace
{
#if defined(__APPLE__)
    constexpr char const* kPlatform = "macos";
#elif defined(__linux__)
    constexpr char const* kPlatform = "linux";
#else
    constexpr char const* kPlatform = "unknown";
#endif

#ifdef NDEBUG
    constexpr char const* kMode = "opt";
#else
    constexpr char const* kMode = "default";
#endif

    constexpr int Spades = 0;
    constexpr int North = 0;
    constexpr int East = 1;
    constexpr int South = 2;
    constexpr int West = 3;

    // Identical to parity_reference.cpp's own make_two_card_finesse_root
    // and exhaustive_layout_source_integration_test.cpp's own -- the third
    // independent copy of this fixture in the tree, matching those two
    // files' own established precedent rather than reaching across a
    // package boundary for it.
    auto make_two_card_finesse_root() -> Deal
    {
        Deal root{};
        root.trump = DDS_NOTRUMP;
        root.first = East;
        root.remainCards[North][Spades] = (1u << 9);
        root.remainCards[South][Spades] = (1u << 8);
        root.remainCards[East][Spades] = (1u << 2) | (1u << 3);
        root.remainCards[West][Spades] = (1u << 4) | (1u << 5) | (1u << 6) | (1u << 7) | (1u << 10);
        return root;
    }

    struct Config
    {
        char const* label;
        be::EvaluateOptions options;
    };

    auto run_config(
        char const* pi_label, be::DeclarerStrategy const& pi, Config const& config, int iterations) -> void
    {
        Deal const root = make_two_card_finesse_root();
        constexpr std::uint64_t seed = 7u;  // parity_reference.cpp's own SAMPLED_SEED

        // Accumulated and printed below (not a per-iteration volatile) so
        // nothing above is optimised away as dead code -- genuinely read,
        // not merely written, which is what makes this survive -Werror's
        // own unused-but-set-variable check as well as an optimiser's.
        // Not a correctness check either way; that is what the gtest/py_test
        // guards are for.
        double sink = 0.0;
        auto const start = std::chrono::steady_clock::now();
        for (int i = 0; i < iterations; ++i)
        {
            be::ExhaustiveLayoutSource const source(root, North, seed);
            be::EvaluationResult const result = be::evaluate(
                root, North, /*tricks_needed=*/1, source, pi, strategies::trivial_defender_play,
                config.options);
            if (! result.error.has_value())
            {
                sink += result.by_strategy.at(pi.id).p_make;
            }
        }
        auto const end = std::chrono::steady_clock::now();
        double const total_ms = std::chrono::duration<double, std::milli>(end - start).count();
        std::printf(
            "cpp pi=%-9s config=%-16s mode=%s platform=%s iterations=%d ms_per_iteration=%.6f sink=%.6f\n",
            pi_label, config.label, kMode, kPlatform, iterations, total_ms / iterations, sink);
    }
}  // namespace

auto main(int argc, char** argv) -> int
{
    int iterations = 2000;  // the earlier order-of-magnitude check's own loop count, for a direct re-run
    if (argc > 1)
    {
        iterations = std::atoi(argv[1]);
    }

    // Same root, same declarer, same tricks_needed, same seed, same
    // sample_size/scan_budget throughout -- only replenish_below differs,
    // so "replenishing" vs "non-replenishing" isolates exactly the
    // node-local replenishment scan's own effect on how often delta is
    // called, per this task's own background.
    be::EvaluateOptions non_replenishing{};
    non_replenishing.sampling.sample_size = 10u;
    non_replenishing.sampling.scan_budget = 100u;

    be::EvaluateOptions replenishing = non_replenishing;
    replenishing.sampling.replenish_below = 5u;  // parity_reference.cpp's own SAMPLED_REPLENISH_BELOW

    Config const configs[] = {
        Config{"non_replenishing", non_replenishing},
        Config{"replenishing", replenishing},
    };

    be::DeclarerStrategy const trivial{.id = 1, .play = strategies::trivial_declarer_play, .state_key = nullptr};
    be::DeclarerStrategy const real_work{
        .id = 2, .play = strategies::real_work_declarer_play, .state_key = nullptr};

    for (Config const& config : configs)
    {
        run_config("trivial", trivial, config, iterations);
        run_config("real_work", real_work, config, iterations);
    }
    return 0;
}

File: benchmarks/python_cost/timer.py
"""The Python side of this capability's own Python-vs-C++ per-callback cost
comparison -- timer.cpp's own header comment explains the shape (two pi,
two sampling configurations) and why. Loops the identical evaluate() call
through the compiled binding and times it.

A py_binary (built, not run automatically -- py_binary is not a test, so
`bazel test //...` never executes it), not a plain script: unlike
benchmarks/belief_evaluation's own sweep scripts, this needs the compiled
belief_space_local_evaluation extension itself, which only bazel's own
py_binary/py_test wiring builds and places on sys.path correctly (see
BUILD.bazel's own comment on this file's target).

Usage (normally driven by report.py, not run standalone):

    bazel run //benchmarks/python_cost:timer_py -- [ITERATIONS]
"""

from __future__ import annotations

import sys
import time

from belief_space_local_evaluation import Card
from belief_space_local_evaluation import ExhaustiveLayoutSource
from belief_space_local_evaluation import evaluate

Spades = 0
North, East, South, West = 0, 1, 2, 3
DDS_NOTRUMP = 4


def make_two_card_finesse_root() -> dict:
    # Identical to parity_reference.cpp's own make_two_card_finesse_root
    # and timer.cpp's own copy -- the fourth independent copy of this tiny
    # fixture in the tree (parity_reference.cpp,
    # exhaustive_layout_source_integration_test.cpp, timer.cpp, this one),
    # matching those files' own established precedent of duplicating a
    # fixture this small rather than sharing it across a package boundary.
    remain_cards = [[0, 0, 0, 0] for _ in range(4)]
    remain_cards[North][Spades] = 1 << 9
    remain_cards[South][Spades] = 1 << 8
    remain_cards[East][Spades] = (1 << 2) | (1 << 3)
    remain_cards[West][Spades] = (1 << 4) | (1 << 5) | (1 << 6) | (1 << 7) | (1 << 10)
    return {
        "trump": DDS_NOTRUMP,
        "first": East,
        "remain_cards": remain_cards,
        "current_trick_suit": (0, 0, 0),
        "current_trick_rank": (0, 0, 0),
    }


def seat_on_play(deal: dict) -> int:
    played = 0
    for rank in deal["current_trick_rank"]:
        if rank == 0:
            break
        played += 1
    return (deal["first"] + played) % 4


def lowest_card_in(remain_cards_row) -> Card:
    for suit in range(4):
        mask = remain_cards_row[suit]
        if mask:
            rank = 2
            while not (mask & (1 << rank)):
                rank += 1
            return Card(suit, rank)
    raise AssertionError("seat holds nothing")


def lowest_legal_card(deal: dict, seat: int) -> Card:
    remain_cards = deal["remain_cards"][seat]
    if deal["current_trick_rank"][0] != 0:
        led = deal["current_trick_suit"][0]
        if remain_cards[led] != 0:
            return lowest_card_in([remain_cards[led] if s == led else 0 for s in range(4)])
    return lowest_card_in(remain_cards)


def trivial_declarer_play(state, view):
    del view
    return lowest_legal_card(state.known_holdings, seat_on_play(state.known_holdings))


def trivial_defender_play(layout, seat, state):
    del state
    return [(lowest_legal_card(layout, seat), 1.0)]


def enumerate_legal_cards(deal: dict, seat: int) -> list[Card]:
    def collect(suit: int) -> list[Card]:
        holding = deal["remain_cards"][seat][suit]
        return [Card(suit, rank) for rank in range(2, 15) if holding & (1 << rank)]

    led = -1
    if deal["current_trick_rank"][0] != 0:
        led = deal["current_trick_suit"][0]
    if led != -1 and deal["remain_cards"][seat][led] != 0:
        return collect(led)
    cards: list[Card] = []
    for suit in range(4):
        cards.extend(collect(suit))
    return cards


def higher_defender_count(layout: dict, declarer: int, suit: int, rank: int) -> int:
    east = (declarer + 1) % 4
    west = (declarer + 3) % 4
    holding = layout["remain_cards"][east][suit] | layout["remain_cards"][west][suit]
    above_rank = holding & ~((1 << (rank + 1)) - 1)
    return bin(above_rank).count("1")


def real_work_declarer_play(state, view):
    # Identical rule to test_support.hpp's safety_score_declarer_play and
    # timer.cpp's own copy -- proven correct (both languages, independently
    # hand-derived) in safety_score_declarer_play_test.cpp and
    # test_belief_space_local_evaluation_python_cost.py; not re-proven
    # here, only timed.
    known_holdings = state.known_holdings
    seat = seat_on_play(known_holdings)
    candidates = enumerate_legal_cards(known_holdings, seat)

    best = None
    best_score = 0.0
    for candidate in candidates:
        score = sum(
            entry.posterior * higher_defender_count(entry.layout, state.declarer, candidate.suit, candidate.rank)
            for entry in view.entries
        )
        if best is None or score < best_score or (score == best_score and candidate.rank < best.rank):
            best = candidate
            best_score = score
    assert best is not None
    return best


def run_config(pi_label: str, pi, config_label: str, kwargs: dict, iterations: int) -> None:
    root = make_two_card_finesse_root()
    seed = 7  # timer.cpp's own seed, parity_reference.cpp's SAMPLED_SEED

    start = time.perf_counter()
    for _ in range(iterations):
        source = ExhaustiveLayoutSource(root, North, seed)
        result = evaluate(root, North, 1, source, pi, trivial_defender_play, **kwargs)
        if "error" not in result:
            _ = result["by_strategy"][1]["p_make"]  # touch the result, nothing more
    total_ms = (time.perf_counter() - start) * 1000.0
    print(
        f"py pi={pi_label:<9s} config={config_label:<16s} iterations={iterations} "
        f"ms_per_iteration={total_ms / iterations:.6f}")


def run(iterations: int) -> None:
    # Same root/declarer/tricks_needed/seed/sample_size/scan_budget
    # throughout -- only replenish_below differs -- matching timer.cpp's
    # own pairing exactly, for the same reason (this task's own
    # background: isolate the node-local replenishment scan's own effect
    # on how often delta is called).
    non_replenishing = {"collect_counters": False, "sample_size": 10, "scan_budget": 100}
    replenishing = dict(non_replenishing, replenish_below=5)

    for config_label, kwargs in (("non_replenishing", non_replenishing), ("replenishing", replenishing)):
        run_config("trivial", trivial_declarer_play, config_label, kwargs, iterations)
        run_config("real_work", real_work_declarer_play, config_label, kwargs, iterations)


if __name__ == "__main__":
    run(int(sys.argv[1]) if len(sys.argv) > 1 else 2000)
